src/transformers/generation/utils.py

```diff
@@ -4915,3 +5242,76 @@ def _concat(data):
 
     # Return a new object of the inferred class with the concatenated attributes
     return model_output_cls(**concatenated_data)
+
+
+def _dola_select_contrast(
+    candidate_premature_layers: List[int],
+    candidate_premature_logits: Dict[int, torch.FloatTensor],
+    final_logits: torch.FloatTensor,
+) -> torch.FloatTensor:
+    if len(candidate_premature_layers) == 1:
+        base_logits = candidate_premature_logits[candidate_premature_layers[0]]
+        final_logits, base_logits = _relative_top_filter(final_logits, base_logits)
+        logits = final_logits - base_logits
+    else:
+        # 1. Stacking all premature_layers into a new dimension
+        stacked_premature_layers = torch.stack(
+            [candidate_premature_logits[i] for i in candidate_premature_layers], dim=0
+        )
+
+        # 2. Calculate the softmax values for mature_layer and all premature_layers
+        softmax_mature_layer = F.softmax(final_logits, dim=-1)  # shape: (batch_size, vocab_size)
+        softmax_premature_layers = F.softmax(
+            stacked_premature_layers, dim=-1
+        )  # shape: (num_premature_layers, batch_size, vocab_size)
+
+        # 3. Calculate M, the average distribution
+        M = 0.5 * (
+            softmax_mature_layer[None, :, :] + softmax_premature_layers
+        )  # shape: (num_premature_layers, batch_size, vocab_size)
+
+        # 4. Calculate log-softmax for the KL divergence
+        log_softmax_mature_layer = F.log_softmax(final_logits, dim=-1)  # shape: (batch_size, vocab_size)
+        log_softmax_premature_layers = F.log_softmax(
+            stacked_premature_layers, dim=-1
+        )  # shape: (num_premature_layers, batch_size, vocab_size)
+
+        # 5. Calculate the KL divergences and then the JS divergences
+        kl1 = F.kl_div(log_softmax_mature_layer[None, :, :], M, reduction="none").mean(
+            -1
+        )  # shape: (num_premature_layers, batch_size)
+        kl2 = F.kl_div(log_softmax_premature_layers, M, reduction="none").mean(
+            -1
+        )  # shape: (num_premature_layers, batch_size)
+        js_divs = 0.5 * (kl1 + kl2)  # shape: (num_premature_layers, batch_size)
+
+        # 6. Reduce the batchmean
+        js_divs = js_divs.mean(-1)  # shape: (num_premature_layers,)
+        premature_layer = candidate_premature_layers[int(js_divs.argmax().cpu().item())]
+
+        base_logits = candidate_premature_logits[premature_layer]
+        final_logits, base_logits = _relative_top_filter(final_logits, base_logits)
+        logits = final_logits - base_logits
+    return logits
+
+
+def _relative_top_filter(
```

Comment:
Definition of objects should go above the lines they're first used

Reply:
Fixed the order!